core/src/main/scala/filodb.core/memstore/OnDemandPagingShard.scala

```diff
@@ -61,35 +61,41 @@ TimeSeriesShard(dataset, storeConfig, shardNum, rawStore, metastore, evictionPol
     val allDataCols = dataset.dataColumns.map(_.id)
 
     // 1. Fetch partitions from memstore
-    val indexIt = iteratePartitions(partMethod, chunkMethod)
-
-    // 2. Determine missing chunks per partition and what to fetch
+    val iterResult = iteratePartitions(partMethod, chunkMethod)
+    val partIdsNotInMemory = iterResult.partIdsNotInMemory
+    // 2. Now determine list of partitions to ODP and the time ranges to ODP
     val partKeyBytesToPage = new ArrayBuffer[Array[Byte]]()
-    val inMemoryPartitions = new ArrayBuffer[ReadablePartition]()
-    val methods = new ArrayBuffer[ChunkScanMethod]
-    indexIt.foreach { p =>
-      chunksToFetch(p, chunkMethod, pagingEnabled).map { rawChunkMethod =>
-        methods += rawChunkMethod   // TODO: really determine range for all partitions
-        partKeyBytesToPage += p.partKeyBytes
-      }.getOrElse {
-        // add it to partitions which do not need to be ODP'ed, send these directly and first
-        inMemoryPartitions += p
+    val pagingMethods = new ArrayBuffer[ChunkScanMethod]
+    val inMemOdp = debox.Set.empty[Int]
+
+    iterResult.partIdsInMemoryMayNeedOdp.foreach { case (pId, startTime) =>
+      val p = partitions.get(pId)
+      if (p != null) {
+        checkIfReallyNeedToOdp(p, chunkMethod, pagingEnabled, startTime).map { rawChunkMethod =>
```

Comment:
Minor point- I think `chunksToFetch` or `chunksToODP` is still a better name for this method.  `checkIf....` seems to imply it returns a Boolean, but it actually returns the ChunkMethod.